### DISCUSSION TITLE
Help with the calculation of substructure of ChEMBL compounds

### DISCUSSION THREAD

**Discussion by mercedesgarnham** [rdkit/rdkit#7848] [Q&A]
We’d like to compare all ChEMBL molecules against one another to identify which ones are substructures. We’re looking to do something like this:
![image](https://github.com/user-attachments/assets/1af8e50f-3ffd-4c42-98be-0cbd2cd8ed5d)
We need the columns "CHEMBLID_1, CHEMBLID_2, Is Substructure?" I put the structures just to illustrate the example. 

We have been using RDKit like this:

mol_CHEMBLID_1 = Chem.MolFromSmiles(smiles_CHEMBLID_1)
mol_CHEMBLID_2 = Chem.MolFromSmiles(smiles_CHEMBLID_2)
rs_1 = mol_CHEMBLID_1.HasSubstructMatch(mol_CHEMBLID_2, useChirality=True)
print(rs_1)
rs_2= mol_CHEMBLID_2HasSubstructMatch(mol_CHEMBLID_1 useChirality=True)
print(rs_2)

The problem with this is that it takes too much time to compare the 2.4 million compounds from ChEMBL. 
Do you have any suggestions on how we can do this more efficiently and fast?

**Comment by bp-kelley** [comment]
You may want to use the SubstructLibrary, this is multi-threaded and can use the pattern fingerprints for fast screen  outs.

http://rdkit.blogspot.com/2018/02/introducing-substructlibrary.html

But you still are going to compare 2.4M compounds against 2.4M compounds...

**Comment by mercedesgarnham** [comment]
Thank you! I've been using that blog entry and this is my current code: 

# Imports
from rdkit import Chem
from rdkit import RDLogger
from rdkit.Chem import rdSubstructLibrary
from rdkit.Chem import rdMolEnumerator
from rdkit.Chem import rdTautomerQuery
from rdkit.Chem import Draw
from rdkit.Chem.Draw import IPythonConsole
IPythonConsole.drawOptions.minFontSize = 10
Draw.SetComicMode(IPythonConsole.drawOptions)
from rdkit.Chem import rdDepictor
rdDepictor.SetPreferCoordGen(True)
import rdkit
print(rdkit.__version__)
import pickle
import gzip
import time
print(time.asctime())

gz = gzip.GzipFile('chembl_34.sdf.gz')

suppl = Chem.ForwardSDMolSupplier(gz)
RDLogger.DisableLog("rdApp.warning")
t1=time.time()
data = []
for i,mol in enumerate(suppl):
    if not ((i+1)%50000):
        print(f"Processed {i+1} molecules in {(time.time()-t1):.1f} seconds")
    if mol is None or mol.GetNumAtoms()>50:
        continue
    fp = Chem.PatternFingerprint(mol,fpSize=1024,tautomerFingerprints=True)
    smi = Chem.MolToSmiles(mol)
    data.append((smi,fp))
t2=time.time()
pickle.dump(data,open('chembl31_sssdata.pkl','wb+'))
t1=time.time()
mols = rdSubstructLibrary.CachedTrustedSmilesMolHolder()
fps = rdSubstructLibrary.TautomerPatternHolder(1024)
for smi,fp in data:
    mols.AddSmiles(smi)
    fps.AddFingerprint(fp)
library = rdSubstructLibrary.SubstructLibrary(mols,fps)
t2=time.time()
print(f"That took {t2-t1:.2f} seconds. The library has {len(library)} molecules.")
pickle.dump(library,open('chembl31_ssslib.pkl','wb+'))

with open('chembl31_ssslib.pkl','rb') as inf:
    sslib = pickle.load(inf)
print(f'SubstructLibrary loaded with {len(sslib)} molecules')

# Prepare to store the results
results = []

# Loop through all molecules in the original data
for i, (smi_a, _) in enumerate(data):
    qry_a = Chem.MolFromSmiles(smi_a)  # Convert SMILES A to a molecule
    bndl_a = rdMolEnumerator.Enumerate(qry_a)  # Generate bond variants of A
    
    for j, (smi_b, _) in enumerate(data):
        if i == j:  # Skip comparing the molecule to itself
            continue

        qry_b = Chem.MolFromSmiles(smi_b)  # Convert SMILES B to a molecule
        bndl_b = rdMolEnumerator.Enumerate(qry_b)  # Generate bond variants of B

        # Check if A is a substructure of B
        matches_a_in_b = sslib.GetMatches(bndl_a)
        is_a_in_b = any(matches_a_in_b)  # True if there are matches
        
        # Check if B is a substructure of A
        matches_b_in_a = sslib.GetMatches(bndl_b)
        is_b_in_a = any(matches_b_in_a)  # True if there are matches
        
        # Append the results in the specified format
        results.append((smi_a, smi_b, 'T' if is_a_in_b else 'F'))
        results.append((smi_b, smi_a, 'F' if is_a_in_b else 'T'))

# Optionally, save the results to a file
with open('substructure_results.txt', 'w') as outfile:
    for mol_a, mol_b, match in results:
        outfile.write(f'{mol_a} {mol_b} {match}\n')

print(f'Comparison results saved with {len(results)} pairs.')

But since this is a for inside an other for it's taking for ever to finish!

**Comment by bp-kelley** [comment]
Ok, I used chembl_downloader to get ChEMBL.smi, 2,409,270 smiles strings in rdkit sanitized format

Using an M1 Max

Making the library takes 20 minutes or so
```
>>> from rdkit.Chem import rdSubstructLibrary
>>> mols = rdSubstructLibrary.CachedTrustedSmilesMolHolder()
>>> fps = rdSubstructLibrary.PatternHolder()
>>> for line in open("chembl.smi"):
...     _=mols.AddSmiles(line.strip()) # _ is so we don't write to the console in interactive mode\
... 
>>> sslib = rdSubstructLibrary.SubstructLibrary(mols)
>>> rdSubstructLibrary.AddPatterns(sslib, fps) # adds the patterns in a multi-threaded way
```

This creates a substructure library with the pattern fps

To search

```
>>> from rdkit import Chem
>>> from tqdm import tqdm
>>> for smi in tqdm(smiles):
...     m = Chem.MolFromSmiles(smi)
...     if not m: continue
...     matches = sslib.GetMatches(m, maxResults=1_000_000)
... 
  0%|                                                                                       | 920/2409270 [00:29<18:32:10, 36.09it/s]
```

Looks to take between half a day and a day to search which doesn't seem so bad.

Note I'm not Saving the matches (they are integer indices, so should be quick to save/store)

Note that the substruct library can be serialized so you can make it once and send it around.

Hope this helps.

**Reply by mercedesgarnham** [reply]
Thank you very much for your response!
I can't find the way to use chembl_downloader to get ChEMBL.smi

**Comment by bp-kelley** [comment]
It's in their readme

https://github.com/cthoyt/chembl-downloader

```
import chembl_downloader

for smiles in chembl_downloader.iterate_smiles():
    print(smiles)
```

Instead of printing, you can write to a file

**Reply by bp-kelley** [reply]
Note that it downloads the chembl DB so you will need free space

**Reply by cthoyt** [reply]
you might want to check https://chembl-downloader.readthedocs.io/en/latest/sdf.html, there's also a function that iterates directly over the parsed rdkit molecules: https://chembl-downloader.readthedocs.io/en/latest/api/chembl_downloader.supplier.html#chembl_downloader.supplier. Example usage:


```python
from rdkit import Chem

import chembl_downloader

data = []
with chembl_downloader.supplier() as suppl:
    for i, mol in enumerate(suppl):
        if mol is None or mol.GetNumAtoms() > 50:
            continue
        fp = Chem.PatternFingerprint(mol, fpSize=1024, tautomerFingerprints=True)
        smi = Chem.MolToSmiles(mol)
        data.append((smi, fp))
```
